Win32-network/src/System/Win32/Async/IOData.hsc

```diff
@@ -0,0 +1,366 @@
+    | ResultSync  !a                     !Bool
+    -- ^ Synchronous result; the if the 'Bool' value is true, `ioDataPtr` will
+    -- be deallocated.
+    | ErrorSync   !(ErrorCode asyncType) !Bool
+    -- ^ Synchronous error; the if the 'Bool' value is true, `ioDataPtr` will
+    -- be deallocated.
+
+
+-- | File descriptors.
+--
+data FD asyncType where
+    FDHandle :: HANDLE -> FD Async
+    FDSocket :: SOCKET -> FD WsaAsync
+
+
+-- | Allocate 'IOCPData' on process heap.  If the continuation does not succeed
+-- and 'getLastError' is not 'ERROR_IO_PENDING' it might deallocate 'IOCPData'
+-- otherwise the resposibility of deallocation is in the
+-- 'dequeueCompletionPackets' thread.
+--
+-- It is executed with async exception masked to ensure that 'ioDataPtr' is
+-- filled in with 'IOCPData' before it is passed to the continuation.
+--
+withIOCPData :: forall a (asyncType :: AsyncType).
+                Storable (IOData asyncType)
+             => String
+             -- ^ error tag
+             -> FD asyncType
+             -- ^ file descriptor used for the syscall
+             -> (Ptr (OverlappedType asyncType)
+                  -> MVar (Either ErrCode Int)
+                  -> IO (Result asyncType a))
+             -- ^ continuation, executed with unmasked async exceptions which
+             -- executes a syscall.  It receives 'LPOVERLAPPED' allocated on
+             -- process heap and an empty 'MVar' which will be filled by
+             -- 'dequeueCompletionPackets' ('IOManager' thread), when the
+             -- async IO terminates.
+             --
+             -- Note that the 'MVar' holds 'ErrorCode Async', even if
+             -- 'asyncType ~ WsaAsync', this is because the
+             -- 'GetQueuedCompletionStatus' and thus
+             -- 'dequeueCompletionPackets' can handle both `windows.h` and
+             -- `winsock2.h` apis.
+             --
+             -> IO a
+withIOCPData errorTag fd k = mask $ \unmask -> do
+    -- both 'IOData' and the stable pointer are allocated here and free either
+    --  * by 'IOManager' thread ('dequeueCompletionPackets')
+    --  * or here in case of some synchronous operations which do not push
+    --    a notifiction to IOCP.
+    -- We cannot simply use `allocaBytes` and make the allocation local, since
+    -- in case of async exception the IOCP thread will access already freed
+    -- memory.
+    --
+    -- TODO: hovewer, if we synchronously cancel operations we could use local
+    -- allocations for both 'IOData' and the stable pointer, which would make
+    -- this library much safer.  For that reason we woudl need to solve two problems:
+    --
+    --   *  'CancelIoEx' does not wait for an operation to be cancelled, we
+    --      would still need to do that.  The simplest solution which could work
+    --      is to await on the 'waitVar' (the thread that was waiting on it was
+    --      cancelled, so we could safely do that).
```

Comment:
I wouldn't bother. The dynamic allocation and deallocation is ok. Sounds like trying to make more things synchronous here would add complexity.